migrate layout from version 2 to 3

akosyakov opened the pull request:
<!--
Thank you for your Pull Request. Please provide a description and review
the requirements below.

Contributors guide: https://github.com/theia-ide/theia/blob/master/CONTRIBUTING.md
-->

#### What it does
<!-- Include relevant issues and describe how they are addressed. -->

fix #5950:
- introduced layout migration points
- implemented migration point for version 2 to discard all unversioned layouts (for backward compatibility)
- implemented scm migration point for version 3 to wrap scm widget into view container
- implemented explorer migration point for version 3 to wrap explorer widget into view container

#### How to test
<!-- Explain how a reviewer can reproduce a bug, test new functionality or verify performance improvements. -->

- get Theia before view containers from fe938acc3aec8fc563bda46bf9c129d9d90c2c32 commit
- open some workspace with it and customize the layout, especially explorer and scm views
- open the same workspace with Theia from this PR, the layout should be restored properly and you should see that some migrations were found in the dev console
- after that reload Theia again to check that Theia from this PR does not apply any migrations to already migrated (latest) layout

#### Review checklist

- [x] as an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

a1994846931931 commented:
It seems that the layout in the bottom area cannot be retrieved.

old-view-1
![old-view](https://user-images.githubusercontent.com/17487291/63234563-55009680-c268-11e9-9da4-ff639ac0d7ef.PNG)

new-view-1
![new-view](https://user-images.githubusercontent.com/17487291/63234565-5762f080-c268-11e9-8108-d59fcfd6d908.PNG)

old-view-2
![old-view-1](https://user-images.githubusercontent.com/17487291/63234566-592cb400-c268-11e9-8553-26c4756a2828.PNG)

old-view-2
![new-view-1](https://user-images.githubusercontent.com/17487291/63234569-5b8f0e00-c268-11e9-94c9-9f8d4d479737.PNG)

akosyakov commented on packages/navigator/src/browser/navigator-frontend-module.ts:
```diff
@@ -65,6 +71,36 @@ export default new ContainerModule(bind => {
             return viewContainer;
         }
     }));
+    bind<ApplicationShellLayoutMigration>(ApplicationShellLayoutMigration).toConstantValue({
```
for each extension? `navigator-migrations.ts` file?

AlexTugarev replied:
that's would be good!

AlexTugarev replied:
or just `layout-migrations.ts` in each extensions.

akosyakov replied:
duplicate file names are bad for search

akosyakov replied:
but `layout-migrations` suffix sounds better

akosyakov commented on packages/scm/src/browser/scm-frontend-module.ts:
```diff
@@ -58,6 +64,36 @@ export default new ContainerModule(bind => {
             return viewContainer;
         }
     })).inSingletonScope();
+    bind<ApplicationShellLayoutMigration>(ApplicationShellLayoutMigration).toConstantValue({
```
ok, will do

akosyakov commented:
thank you for reviewing guys, will have a look what's wrong with the bottom panel

it should not have been affected by view containers 🤔

AlexTugarev commented:
Might be unrelated to view containers, but to the layout migrator:

`WARN Couldn't restore widget for problem. Error: Error: No widget factory 'problem' has been registered. `

Could this cause the bottom panel to be restored "unfinished". I see there is not exception handling for migrations, is that by intention?

akosyakov commented:
> Could this cause the bottom panel to be restored "unfinished". I see there is not exception handling for migrations, is that by intention?

Yes, i wanted them to fail, it does not make sense to apply other migrations if first misbehaves, since later depends on former. Should i put it in a comment? Makes sense?

But i don't think migrations related to problems at all. From warning sounds like it was always there. Good that we found it now.

akosyakov commented:
@marechal-p @AlexTugarev @a1994846931931 I've addressed all your comments, please have a look.

Also you have really sharp eyes to find an issue with problems. I have not even noticed that i broke it 🙈

AlexTugarev commented:
Testing right now...

AlexTugarev commented:
![Screen Shot 2019-08-21 at 13 20 15](https://user-images.githubusercontent.com/914497/63427850-70110900-c416-11e9-9ac4-6a4768cd5cee.png)

> Found 3 migrations from layout version 2 to version 3, migrating...

Works fine now as described in "How to test"! 🎉

AlexTugarev commented on packages/core/src/browser/shell/shell-layout-restorer.ts:
```diff
@@ -147,30 +209,68 @@ export class ShellLayoutRestorer implements CommandContribution {
+            } else {
+                console.warn(`Layout version ${layoutVersion} is ahead current layout version ${applicationShellLayoutVersion}, trying to load anyway...`);
+            }
+            console.info(`Please use '${resetLayoutCommand.label}' command if the layout looks bogus.`);
```
This is not really visible to users, isn't it?

Do we know if a migration failed? If so, wdyt of notifying with the option to trigger the layout reset?

akosyakov replied:
Do you mean it should be `console.log`? 

I can remove it, i've added just for a case if layout does not look ok, but migrations did not fail or layout restoration, for example if migrations are actually missing, and a user had a look at console and get a hint how to recover.

AlexTugarev replied:
> Do we know if a migration failed?

that's what I mean.

it would be a bad UX if something happen there and we cannot notify the user (via notification.) the issue we've seen when testing this PR, shows that a migration may break the layout somehow. do we really expect users to open JS console to inspect the logs? if possible I'd rather show a message and offer an option to resolve the situation. if that's out of scope here, I'd propose to tackle this separately.

akosyakov replied:
If a migration failed then layout will be completely reset. Do you want me to show a user notification as well?

If migrations are not failed, but layout is bogus, then it is a programming error.

AlexTugarev replied:
> If migrations are not failed, but layout is bogus, ...

that we cannot detect? is that correct?

akosyakov replied:
> that we cannot detect? is that correct?

yes, i mean that we just missed something, like i've missed `problems` widget. It was small thing this time, another time a widget can be completely unusable. We can detect it only by testing and fix by adding migrations.
